### PR TITLE
feat: [GH-1194]: support users with viewer roles to be able to mutate the dashboard variable values to customise views

### PR DESCRIPTION
**Pull request by Vikrant2520**
### Summary

- support local storage support for the dashboard to localize the views for the dashboard 
- this would remove the dashboard update on API support and support viewer roles to view different views without updating the dashboard. 
- Helps us to keep track of dashboard state for multiple dashboards and persist the same across user sessions 

#### Related Issues / PR's

https://github.com/SigNoz/engineering-pod/issues/1194

#### Screenshots

https://github.com/SigNoz/signoz/assets/54737045/d6c966f7-3a8c-46f2-a970-441acca81811

#### Affected Areas and Manually Tested Areas

- Dashboard variables / dependent dashboard variables

## Summary by CodeRabbit

- **New Features**
	- Introduced a new icon for better visual representation in the `DynamicColumnTable`.
	- Added `DASHBOARD_VARIABLES` to manage dashboard-specific settings more efficiently.
	- Enhanced the dashboard's ability to retain user-specific variables through local storage for a personalized experience.

- **Refactor**
	- Improved `DynamicColumnTable` styling for better layout and responsiveness across devices.
	- Streamlined dashboard variable management with new and updated hooks for better performance and usability.

- **Style**
	- Updated button styling within the `DynamicColumnTable` for a more cohesive look and feel.

### CODE REVIEW

File: frontend/src/container/NewDashboard/DashboardVariablesSelection/DashboardVariableSelection.tsx [hunk 1]
```diff
@@ -1,18 +1,20 @@
...
 
 function DashboardVariableSelection(): JSX.Element | null {
-	const { selectedDashboard, setSelectedDashboard } = useDashboard();
+	const {
+		selectedDashboard,
+		setSelectedDashboard,
+		dashboardId,
+	} = useDashboard();
+
+	const { updateLocalStorage } = useDashboardFromLocalStorage(dashboardId);
```
Comment (YounixM) [inline frontend/src/container/NewDashboard/DashboardVariablesSelection/DashboardVariableSelection.tsx:17]: naming can be better.

File: frontend/src/hooks/dashboard/useDashboardFromLocalStorage.tsx [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		allSelected: boolean,
+	) => void;
+}
+
+export const useDashboardFromLocalStorage = (
+	dashboardId: string,
+): UseDashboardFromLocalStorageReturn => {
+	const [allDashboards, setAllDashboards] = useState<DashboardLocalStorage>({});
+
+	const [currentDashboard, setCurrentDashboard] = useState<LocalStoreDashboard>(
```
Comment (YounixM) [inline frontend/src/hooks/dashboard/useDashboardFromLocalStorage.tsx:32]: We could have made these simple util functions that update the value. We can avoid storing context for the same.

Comment (Vikrant2520) [reply frontend/src/hooks/dashboard/useDashboardFromLocalStorage.tsx:32]: we used custom hook for the same :)

File: frontend/src/container/NewDashboard/DashboardVariablesSelection/DashboardVariableSelection.tsx (hunk 1 above)
Comment (Vikrant2520) [reply frontend/src/container/NewDashboard/DashboardVariablesSelection/DashboardVariableSelection.tsx]: updated

**Review by YounixM — APPROVED** [review]

**Review by Rajat-Dabade — APPROVED** [review]

**Pull request merged**